tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11238,1067 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+}
+
+TEST_F(ExecutionTest, LongVector_Clamp_uint64) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+    WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+  LongVectorOpTestBase<uint64_t>(LongVectorOpType_Clamp);
+}
+
+TEST_F(ExecutionTest, LongVector_Initialize_uint64) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+    WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+  LongVectorOpTestBase<uint64_t>(LongVectorOpType_Initialize);
+}
+
+template <typename T>
+void ExecutionTest::LongVectorOpTestBase(LongVectorOpType opType) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+    WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+
+  LongVectorOpTestConfig<T> TestConfig(opType);
+
+  LongVectorOpTestBase<T, 3>(TestConfig);
+  LongVectorOpTestBase<T, 4>(TestConfig);
+  LongVectorOpTestBase<T, 5>(TestConfig);
+  LongVectorOpTestBase<T, 16>(TestConfig);
+  LongVectorOpTestBase<T, 17>(TestConfig);
+  LongVectorOpTestBase<T, 35>(TestConfig);
+  // TODOLONGVEC: 100, 1024 breaks the size limit for structured buffers
+  //LongVectorOpTestBase<T, 100>(TestConfig);
+  //LongVectorOpTestBase<T, 1024>(TestConfig);
+}
+
+template <typename T, std::size_t N>
+void ExecutionTest::LongVectorOpTestBase(LongVectorOpTestConfig<T>& TestConfig) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+    WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+
+  LogCommentFmt(L"Running LongVectorOpTestBase<%S, %zu>", typeid(T).name(), N);
+
+  CComPtr<ID3D12Device> D3DDevice;
+  if ((!m_HLKModeEnabled && !m_ExperimentalModeEnabled)
+      && !CreateDevice(&D3DDevice, D3D_SHADER_MODEL_6_9)) {
+    WEX::Logging::Log::Comment("Device does not support SM 6.9. Can't run these tests.");
+    WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+    return;
+  }
+
+  DeterministicNumberGenerator<T> NumberGenerator(1337);
+  std::vector<T> VecInput1 = {};
+  std::vector<T> VecInput2 = {};
+  const T ScalarInput = NumberGenerator.generate();
+  const bool IsVectorBinaryOp = TestConfig.IsBinaryOp && !TestConfig.IsScalarOp;
+
+  // Fill the vector inputs with values.
+  for(size_t i = 0; i < N; i++)
+  {
+    // Always generate input. Duh.
+    VecInput1.push_back(NumberGenerator.generate());
+
+    if(IsVectorBinaryOp)
+    {
+      VecInput2.push_back(NumberGenerator.generate());
+    }
+  }
+
+  // We pass these values into the shader and they're requried to compile. So
+  // they need to set to something.
+  T ClampArgC = 0;
+  T ClampArgT = 0;
+  if(TestConfig.IntrinsicString == "testClamp")
+  {
+    ClampArgC = NumberGenerator.generate();
+    ClampArgT = NumberGenerator.generate();
+    while(ClampArgC >= ClampArgT)
+    {
+      // Generate a new value for ClampArgC. It needs to be smaller than
+      // ClampArgT.
+      ClampArgT = NumberGenerator.generate();
+    }
+  }
+
+  std::vector<T> VecExpected = {};
+  for (size_t Index = 0; Index < VecInput1.size(); Index++) {
```

Comment:
This logic is working for now. But I was going to discuss further with the team about adding logic to compute the scalarized expected values in the shader or add reasonable tolerance values for comparing expected.

Reply:
Discussed this, adding tolerances is acceptable. But, the tests are passing now. I will add additional tolerance logic in a subsequent change when adding the rest of the coverage.